### PR TITLE
EWC

### PR DESCRIPTION
**Pull request by BartekCupial**

### CODE REVIEW

File: experiment_code/hackrl/experiment.py [hunk 1]
```diff
@@ -1105,6 +1208,19 @@ def signal_handler(signum, frame):
...
+        global EWC_INSTANCE
+        ttyrec_data = TTYREC_ENVPOOL.result()
+        # idx = TTYREC_ENVPOOL.idx
+        ttyrec_data = TTYREC_ENVPOOL.result()
+        idx = TTYREC_ENVPOOL.idx
+        ttyrec_predictions, TTYREC_HIDDEN_STATE[idx] = model(
+            ttyrec_data, TTYREC_HIDDEN_STATE[idx]
+        )
+
+        EWC_INSTANCE = EWC(learner_state.model, TTYREC_ENVPOOL.result())
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:1222]: I think we want to use more data than one batch. It would be best if you would just pass `TTYREC_ENVPOOL` to `Ewc` and inside called `TTYREC_ENVPOOL.result` few times (or more) with for loop.

File: experiment_code/hackrl/experiment.py [hunk 2]
```diff
@@ -804,6 +903,9 @@ def compute_gradients(data, learner_state, stats):
 
         stats["kickstarting_loss_bc"] += kickstarting_loss_bc.item()
         stats["kickstarting_coeff_bc"] += FLAGS.kickstarting_loss_bc
+    if FLAGS.use_ewc:
+        ewc_penalty = EWC_INSTANCE.penalty(model)
+        total_loss += ewc_penalty * FLAGS.ewc_penalty_scaler
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:908]: It would be great if we logged ewc_penalty similar to other auxiliary losses. Please have a look at `if FLAGS.use_kickstaring` for an example.

File: experiment_code/hackrl/experiment.py [hunk 3]
```diff
@@ -35,18 +39,106 @@
...
+            self._means[n] = p.data
+
+    def _diag_fisher(self):
+        precision_matrices = {}
+        for n, p in deepcopy(self.params).items():
+            p.data.zero_()
+            precision_matrices[n] = p.data
+
+        self.model.eval()
+        cudnn.enabled = False
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:77]: I don't follow why we need this line. Can you please explain?

File: experiment_code/hackrl/experiment.py (hunk 1 above)
Comment (BartekCupial) [reply experiment_code/hackrl/experiment.py:1222]: On average each trajectory in dataset has 10000 steps. Assuming that we need around 300 batches just to ensure we represent whole trajectories with this loss. I'd say we need around 1000 batches for representative sample size.

Comment (BartekCupial) [reply experiment_code/hackrl/experiment.py:1222]: Maybe it's worth to precompute this?

File: experiment_code/hackrl/experiment.py [hunk 4]
```diff
@@ -642,6 +736,10 @@ def compute_gradients(data, learner_state, stats):
 
     total_loss = 0
 
+    if EWC_INSTANCE is None:
+        # It means that we do not load TTYREC
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:740]: What about stricter policy. 
```
assert not (FLAGS.supervised_loss or FLAGS.behavioural_clone)
```

File: experiment_code/hackrl/experiment.py (hunk 3 above)
Comment (MichalBortkiewicz) [reply experiment_code/hackrl/experiment.py:77]: Without this line there is an Error with backpropagation in eval mode for model with RNN
